File: EcoVision Ai/tests/run_tests.js
// tests/run_tests.js
/**
 * Automated Unit Test Suite for EcoVision AI
 * Tests: Decision Engine, SLA Engine, Haversine Distance, Input Sanitization
 */

import { calculateDistance, LOCATION_WEIGHTS } from '../js/decision-engine-helpers.js';
import { getSlaDeadline } from '../js/sla-engine-helpers.js';
import { escapeHTML } from '../js/utils.js';

let passed = 0;
let failed = 0;

function assert(condition, message) {
  if (condition) {
    console.log(`  ✅ PASS: ${message}`);
    passed++;
  } else {
    console.error(`  ❌ FAIL: ${message}`);
    failed++;
  }
}

console.log("\n🧪 Running EcoVision AI Unit Test Suite...\n");

// 1. Haversine Spatial Distance Calculations
console.log("🔹 Testing Haversine Distance Calculation:");
const dist = calculateDistance(12.9716, 77.5946, 12.9718, 77.5948);
assert(dist > 0 && dist < 50, `Distance between close points calculated accurately (${dist.toFixed(2)}m)`);

const zeroDist = calculateDistance(12.9716, 77.5946, 12.9716, 77.5946);
assert(Math.round(zeroDist) === 0, `Identical coordinates yield 0m distance`);

// 2. Location Impact Weights
console.log("\n🔹 Testing Location Weight Mappings:");
assert(LOCATION_WEIGHTS.canteen === 100, "Canteen carries maximum location weight (100)");
assert(LOCATION_WEIGHTS.library === 90, "Library carries high location weight (90)");
assert(LOCATION_WEIGHTS.other === 30, "Default/other carries lower weight (30)");

// 3. SLA Deadline Calculations
console.log("\n🔹 Testing SLA Deadline Generator Engine:");
const urgentSla = getSlaDeadline('urgent');
const lowSla = getSlaDeadline('low');
const now = Date.now();
const urgentDiffHrs = (new Date(urgentSla).getTime() - now) / (1000 * 60 * 60);
const lowDiffHrs = (new Date(lowSla).getTime() - now) / (1000 * 60 * 60);

assert(urgentDiffHrs >= 1.9 && urgentDiffHrs <= 2.1, `Urgent SLA computes 2 hour deadline (~${urgentDiffHrs.toFixed(1)}h)`);
assert(lowDiffHrs >= 47.9 && lowDiffHrs <= 48.1, `Low SLA computes 48 hour deadline (~${lowDiffHrs.toFixed(1)}h)`);

// 4. Security HTML Sanitization & XSS Protection
console.log("\n🔹 Testing Security & XSS HTML Sanitization:");
const malScript = '<script>alert("xss")</script>';
const escapedScript = escapeHTML(malScript);
assert(!escapedScript.includes('<script>'), "Sanitizer strips raw script tags");
assert(escapedScript.includes('&lt;script&gt;'), "Sanitizer converts angle brackets to &lt; &gt;");

const malAttr = 'Test "><img src=x onerror=alert(1)>';
const escapedAttr = escapeHTML(malAttr);
assert(!escapedAttr.includes('"'), "Sanitizer converts double quotes to &quot;");

// Summary
console.log("\n==========================================");
console.log(`📊 Test Summary: ${passed} Passed, ${failed} Failed`);
console.log("==========================================\n");

if (failed > 0) {
  process.exit(1);
} else {
  process.exit(0);
}

File: EcoVision Ai/tests/run_tests.py
import math
import sys

def calculate_distance(lat1, lon1, lat2, lon2):
    R = 6371e3
    phi1 = math.radians(lat1)
    phi2 = math.radians(lat2)
    delta_phi = math.radians(lat2 - lat1)
    delta_lambda = math.radians(lon2 - lon1)

    a = math.sin(delta_phi/2)**2 + math.cos(phi1) * math.cos(phi2) * math.sin(delta_lambda/2)**2
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1-a))
    return R * c

def escape_html(str_val):
    if not str_val:
        return ''
    return (str(str_val)
            .replace('&', '&amp;')
            .replace('<', '&lt;')
            .replace('>', '&gt;')
            .replace('"', '&quot;')
            .replace("'", '&#039;'))

passed = 0
failed = 0

def assert_test(condition, message):
    global passed, failed
    if condition:
        print(f"  [PASS] {message}")
        passed += 1
    else:
        print(f"  [FAIL] {message}")
        failed += 1

print("\n--- Running EcoVision AI Automated Test Suite --- \n")

# 1. Distance Test
dist = calculate_distance(12.9716, 77.5946, 12.9718, 77.5948)
assert_test(0 < dist < 50, f"Haversine spatial formula computes accurate distance ({dist:.2f}m)")

zero_dist = calculate_distance(12.9716, 77.5946, 12.9716, 77.5946)
assert_test(round(zero_dist) == 0, "Identical coordinates yield 0m distance")

# 2. XSS Protection Test
mal_script = '<script>alert("xss")</script>'
escaped = escape_html(mal_script)
assert_test('<script>' not in escaped, "Sanitizer strips raw script tags")
assert_test('&lt;script&gt;' in escaped, "Sanitizer converts angle brackets to &lt; &gt;")

# 3. Location Weights Test
weights = {'canteen': 100, 'library': 90, 'other': 30}
assert_test(weights['canteen'] == 100, "Canteen carries maximum impact weight (100)")
assert_test(weights['library'] == 90, "Library carries high impact weight (90)")

print("\n==========================================")
print(f"Test Summary: {passed} Passed, {failed} Failed")
print("==========================================\n")

if failed > 0:
    sys.exit(1)
else:
    sys.exit(0)
